### PR TITLE
[IE Common][VPU] Fix small input size inference for dynamic model

### PR DESCRIPTION
**Pull request by AndrewBakalinIntel**
Ticket - #-44360
Changes:
- eliminate `Unsqueeze+Gather` pair, when `Gather` gathers data by `1` dimension which was previously added by `Unsqueeze` which is actually doing nothing.
- calculate `K` only once in `StaticShapeTopK`. The problem happens when we have `ShapeOf->Concat->ReduceMin` subgraph for `K` evaluation. If we have pretty small input size, the value that we received from `ShapeOf` may be less that one that it is concatenated with (e.g. ShapeOf `283` vs const `300`), so ReduceMin returns `283`. After `ShapeOf` elimination we don't have a chance to propagate `283` so we get `300` as a result and shape inference fail then. There are no problems with bigger input sizes just because `ShapeOf` always propagates value `>300` and there are no such mismatch.

### CODE REVIEW

**Comment by AndrewBakalinIntel** [comment]
@Maxim-Doronin @andrejsokolov could you please take a look?

**Comment by AndrewBakalinIntel** [comment]
@Maxim-Doronin @andrejsokolov ping.

File: inference-engine/src/transformations/src/transformations/common_optimizations/eliminate_unsqueeze_gather.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#include "transformations/common_optimizations/eliminate_unsqueeze_gather.hpp"
+
+#include <ngraph/opsets/opset6.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/rt_info.hpp>
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::EliminateUnsqueezeGather, "EliminateUnsqueezeGather", 0);
+
+ngraph::pass::EliminateUnsqueezeGather::EliminateUnsqueezeGather() {
+    const auto unsqueezeAxis = ngraph::pattern::any_input();
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/common_optimizations/eliminate_unsqueeze_gather.cpp:14]: Please add `MATCHER_SCOPE` marco. It is needed for conditional compilation feature.
Examples you can find in the PR https://github.com/openvinotoolkit/openvino/pull/3795

File: inference-engine/src/transformations/src/transformations/common_optimizations/eliminate_unsqueeze_gather.cpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        const auto& gatherAxisNode = ngraph::as_type_ptr<ngraph::opset6::Constant>(m_gatherAxis.get_node_shared_ptr());
+
+        if (!unsqueezeAxisNode || !gatherAxisNode) {
+            return false;
+        }
+
+        const auto& unsqueezeAxisVec = unsqueezeAxisNode->cast_vector<int64_t>();
+        const auto& gatherAxisVec = gatherAxisNode->cast_vector<int64_t>();
+
+        if (unsqueezeAxisVec.size() != 1) {
```
Comment (andrejsokolov) [inline inference-engine/src/transformations/src/transformations/common_optimizations/eliminate_unsqueeze_gather.cpp:36]: `gatherAxisVec.size()` should not be checked for == `1`?

File: inference-engine/src/transformations/src/transformations/common_optimizations/eliminate_unsqueeze_gather.cpp (hunk 1 above)
Comment (AndrewBakalinIntel) [reply inference-engine/src/transformations/src/transformations/common_optimizations/eliminate_unsqueeze_gather.cpp:16]: Thanks! Done.

File: inference-engine/src/transformations/src/transformations/common_optimizations/eliminate_unsqueeze_gather.cpp (hunk 2 above)
Comment (AndrewBakalinIntel) [reply inference-engine/src/transformations/src/transformations/common_optimizations/eliminate_unsqueeze_gather.cpp]: Good catch! Fixed.

**Review by andrejsokolov — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

**Review by Maxim-Doronin — APPROVED** [review]

**Comment by AndrewBakalinIntel** [comment]
@Maxim-Doronin could you please merge?

File: inference-engine/src/transformations/src/transformations/common_optimizations/eliminate_unsqueeze_gather.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        const auto& gatherAxisNode = ngraph::as_type_ptr<ngraph::opset6::Constant>(m_gatherAxis.get_node_shared_ptr());
+
+        if (!unsqueezeAxisNode || !gatherAxisNode) {
+            return false;
+        }
+
+        const auto& unsqueezeAxisVec = unsqueezeAxisNode->cast_vector<int64_t>();
+        const auto& gatherAxisVec = gatherAxisNode->cast_vector<int64_t>();
+
+        if (unsqueezeAxisVec.size() != 1 || gatherAxisVec.size() != 1) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/eliminate_unsqueeze_gather.cpp:38]: Please add comments that explains the situation where we can remove this unsqueeze-gather

File: inference-engine/src/transformations/src/transformations/common_optimizations/eliminate_unsqueeze_gather.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            return false;
+        }
+
+        if (unsqueezeAxisVec.front() != gatherAxisVec.front()) {
+            return false;
+        }
+
+        auto& m_gather = patternValue.at(gather);
+        const auto& m_unsqueeze = patternValue.at(unsqueeze);
+        const auto& unsqueezeData = m_unsqueeze.get_node_shared_ptr()->get_input_node_shared_ptr(0);
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/eliminate_unsqueeze_gather.cpp:48]: You can't use `get_node_shared_ptr` and then pass `unsqueezeData` as input. So you have to use output port that comes as input to unsqueeze. And the better way is to take it from predicate.

**Comment by AndrewBakalinIntel** [comment]
Fixed the comments above in https://github.com/openvinotoolkit/openvino/pull/4013

**Pull request merged**